horovod/common/operations.cc

```diff
@@ -1545,15 +1396,16 @@ bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx,
     // ranks at this tick.
 
     // 1. Get message lengths from every rank.
-    auto recvcounts = new int[state.size];
+    auto recvcounts = new int[global_size];
```

Comment:
I think having the Controller be an abstraction over collective ops like Reduce and Gather is too granular.  Ideally, the controller should encapsulate the logic of Parts 1, 2, 3, and 5 here are logically doing one thing: collecting the tensors that are ready from other ranks.  So my suggestion would be to put all of this behind a single method in MPIController called `ReceiveReadyTensors()` or similar.  What do you think?

Reply:
Yes. That would definitely be a further step to extend the controller. Since it does not relate to MPI specific context, I think we can put it at the generic controller level. Does that make sense to you?

Reply:
Done.